kernel/nothread.c

```diff
@@ -37,12 +37,14 @@ int32_t z_impl_k_sleep(k_timeout_t timeout)
 
 	ticks = timeout.ticks;
 	if (Z_TICK_ABS(ticks) <= 0) {
-		expected_wakeup_ticks = ticks + sys_clock_tick_get_32();
+		/* ticks is delta timeout */
+		ticks_to_wait = ticks;
 	} else {
-		expected_wakeup_ticks = Z_TICK_ABS(ticks);
+		/* ticks is absolute timeout expiration */
+		ticks_to_wait = Z_TICK_ABS(ticks) - sys_clock_tick_get_32();
```

Comment:
If we have already passed the absolute point in time (Z_TICKS_ABS(ticks) < sys_clock_tick_get_32()), we should set `ticks_to_wait` to 0, otherwise we are waiting when we should not be waiting.